src/core/thread/energy_scan_server.cpp

```diff
@@ -166,42 +180,35 @@ void EnergyScanServer::HandleScanResult(Mac::EnergyScanResult *aResult)
 
 void EnergyScanServer::SendReport(void)
 {
-    Error                   error = kErrorNone;
-    MeshCoP::ChannelMaskTlv channelMask;
-    MeshCoP::EnergyListTlv  energyList;
-    Tmf::MessageInfo        messageInfo(GetInstance());
-    Coap::Message *         message;
-
-    message = Get<Tmf::Agent>().NewPriorityConfirmablePostMessage(UriPath::kEnergyReport);
-    VerifyOrExit(message != nullptr, error = kErrorNoBufs);
+    Error            error = kErrorNone;
+    Tmf::MessageInfo messageInfo(GetInstance());
+    uint16_t         tlvLength;
 
-    channelMask.Init();
-    channelMask.SetChannelMask(mChannelMask);
-    SuccessOrExit(error = channelMask.AppendTo(*message));
+    // Update the Energy List TLV length in Report message
 
-    energyList.Init();
-    energyList.SetLength(mScanResultsLength);
-    SuccessOrExit(error = message->Append(energyList));
-    SuccessOrExit(error = message->AppendBytes(mScanResults, mScanResultsLength));
+    tlvLength = mReportMessage->GetLength() - mEnergyListTlvOffset - sizeof(Tlv);
+    VerifyOrExit(tlvLength <= NumericLimits<uint8_t>::kMax);
```

Comment:
Should we apply this limit in `HandleScanResult()`?

Reply:
Good idea. Moves it to `HandleScanResult()` in new push. Thanks.